cpp/src/interop/to_arrow_device.cu

```diff
@@ -0,0 +1,590 @@
+                                                            rmm::mr::device_memory_resource* mr,
+                                                            struct ArrowArray* out)
+{
+  CUDF_FAIL("Unsupported type for to_arrow_device");
+}
+
+template <>
+int dispatch_to_arrow_device::operator()<cudf::list_view>(cudf::column& column,
+                                                          cudf::type_id id,
+                                                          rmm::cuda_stream_view stream,
+                                                          rmm::mr::device_memory_resource* mr,
+                                                          struct ArrowArray* out)
+{
+  CUDF_FAIL("Unsupported type for to_arrow_device");
+}
+
+template <>
+int dispatch_to_arrow_device::operator()<cudf::dictionary32>(cudf::column& column,
+                                                             cudf::type_id id,
+                                                             rmm::cuda_stream_view stream,
+                                                             rmm::mr::device_memory_resource* mr,
+                                                             struct ArrowArray* out)
+{
+  CUDF_FAIL("Unsupported type for to_arrow_device");
+}
+
+struct ArrowDeviceArrayPrivateData {
+  struct ArrowArray parent;
+  cudaEvent_t sync_event;
+};
+
+static void ArrowDeviceArrayRelease(struct ArrowArray* array)
+{
+  auto private_data = reinterpret_cast<ArrowDeviceArrayPrivateData*>(array->private_data);
+  cudaEventDestroy(private_data->sync_event);
+  ArrowArrayRelease(&private_data->parent);
+  delete private_data;
+  array->release = nullptr;
+}
+
+}  // namespace
+}  // namespace detail
+
+nanoarrow::UniqueSchema to_arrow_schema(cudf::table_view const& input,
+                                        std::vector<column_metadata> const& metadata)
+{
+  CUDF_EXPECTS((metadata.size() == static_cast<std::size_t>(input.num_columns())),
+               "columns' metadata should be equal to the number of columns in table");
+
+  nanoarrow::UniqueSchema result;
+  ArrowSchemaInit(result.get());
+  if (ArrowSchemaSetTypeStruct(result.get(), input.num_columns()) != NANOARROW_OK) {
+    CUDF_FAIL("failed to initialize ArrowSchema children");
+  }
+
+  // what i would give for a zip iterator....
+  for (int i = 0; i < input.num_columns(); ++i) {
+    auto child = result->children[i];
+    auto col   = input.column(i);
+    ArrowSchemaSetName(child, metadata[i].name.c_str());
+    child->flags = col.nullable() ? ARROW_FLAG_NULLABLE : 0;
+
+    if (col.type().id() == cudf::type_id::EMPTY) {
+      ArrowSchemaSetType(child, NANOARROW_TYPE_NA);
+      continue;
+    }
+
+    auto status = cudf::type_dispatcher(
+      col.type(), detail::dispatch_to_arrow_type{}, col, col.type().id(), metadata[i], child);
+    CUDF_EXPECTS(status == NANOARROW_OK, "type dispatcher failed to determine column type");
```

Comment:
Good point, i'll update that

Reply:
Ah, right now we're only using `ArrowErrorCode` (the integer) we aren't passing around an `ArrowError` to capture the error message. Since it looks like the existing code uses `CUDF_EXPECTS for error handling which throws, we could use `NANOARROW_THROW_NOT_OK` instead of `NANOARROW_RETURN_NOT_OK` and just have it throw on error like the existing code

Reply:
Yup I think this makes sense.